File: backend-old/src/tags/conditional.py
import re
from fastapi import HTTPException
from src.tags.base import BaseTag, TagParseResult


class ConditionalTag(BaseTag):
    """
    Handles [if expr], [elif expr], [else], [/if].

    Opening pushes an 'if_builder' scope. Each elif/else packages the
    previous branch and starts a new one. Closing packages the last
    branch and emits a single 'conditional_block' step.

    Conditions are bare Python expressions (no [python "..."] wrapper).
    The runtime evaluates them with eval() inside the session env.
    """
    name = "conditional"

    IF = re.compile(r'^\[if\s+(.+)\]$')
    ELIF = re.compile(r'^\[(?:elif|else\s+if)\s+(.+)\]$')
    ELSE = re.compile(r'^\[else\]$')
    CLOSE = re.compile(r'^\[/if\]$')

    def parse(self, line, line_idx, ctx):
        stack = ctx["scope_stack"]

        m = self.IF.match(line)
        if m:
            return TagParseResult(
                consumed=True,
                scope_open={
                    "type": "if_builder",
                    "branches": [],
                    "current_branch_type": "if",
                    "current_branch_expr": m.group(1).strip(),
                    "current_branch_steps": [],
                },
            )

        m = self.ELIF.match(line)
        if m:
            if not stack or stack[-1]["type"] != "if_builder":
                raise ValueError(
                    f"Syntax Error line {line_idx + 1}: "
                    f"Unexpected [elif] without an open [if]."
                )
            builder = stack[-1]
            if builder["current_branch_type"] == "else":
                raise ValueError(
                    f"Syntax Error line {line_idx + 1}: "
                    f"[elif] cannot come after [else]."
                )
            builder["branches"].append({
                "mode": builder["current_branch_type"],
                "condition": builder["current_branch_expr"],
                "steps": builder["current_branch_steps"],
            })
            builder["current_branch_type"] = "elif"
            builder["current_branch_expr"] = m.group(1).strip()
            builder["current_branch_steps"] = []
            return TagParseResult(consumed=True)

        if self.ELSE.match(line):
            if not stack or stack[-1]["type"] != "if_builder":
                raise ValueError(
                    f"Syntax Error line {line_idx + 1}: "
                    f"Unexpected [else] without an open [if]."
                )
            builder = stack[-1]
            if builder["current_branch_type"] == "else":
                raise ValueError(
                    f"Syntax Error line {line_idx + 1}: "
                    f"Duplicate [else] in a single conditional block."
                )
            builder["branches"].append({
                "mode": builder["current_branch_type"],
                "condition": builder["current_branch_expr"],
                "steps": builder["current_branch_steps"],
            })
            builder["current_branch_type"] = "else"
            builder["current_branch_expr"] = None
            builder["current_branch_steps"] = []
            return TagParseResult(consumed=True)

        if self.CLOSE.match(line):
            if not stack or stack[-1]["type"] != "if_builder":
                raise ValueError(
                    f"Syntax Error line {line_idx + 1}: Mismatched [/if]."
                )
            return TagParseResult(consumed=True, scope_close="if_builder")

        return TagParseResult(consumed=False)

    def execute(self, step, ctx):
        if step.get("type") != "conditional_block":
            return None

        env = ctx["env"]
        selected = None

        for branch in step["branches"]:
            if branch["mode"] in ("if", "elif"):
                try:
                    result = bool(eval(branch["condition"], {}, env))
                except Exception as e:
                    raise HTTPException(
                        status_code=422,
                        detail={
                            "status": "CONDITIONAL_EVAL_ERROR",
                            "message": f"Failed to evaluate condition: {branch['condition']}",
                            "details": str(e),
                        },
                    )
            else:
                result = True

            if result:
                selected = branch
                break

        if selected:
            return ("inject", selected["steps"])
        return None

File: frontend/src/lib/server/tags/conditional.ts
import { BaseTag, TagParseResult, type ExecResult } from './base.js';
import type { Step, ExecContext, ParseContext } from '../types.js';

interface Branch {
    mode: 'if' | 'elif' | 'else';
    condition: string | null;
    steps: Step[];
}

/**
 * Handles [if expr], [elif expr], [else], [/if].
 *
 * Conditions are evaluated as JavaScript expressions against the
 * shared runtime env. The distinction between Python and TS
 * expressions is not made here — see the Wiki for the rationale.
 */
export class ConditionalTag extends BaseTag {
    name = 'conditional';

    private readonly IF = /^\[if\s+(.+)\]$/;
    private readonly ELIF = /^\[(?:elif|else\s+if)\s+(.+)\]$/;
    private readonly ELSE = /^\[else\]$/;
    private readonly CLOSE = /^\[\/if\]$/;

    parse(line: string, lineIdx: number, ctx: ParseContext): TagParseResult {
        const stack = ctx.scope_stack;

        const ifm = this.IF.exec(line);
        if (ifm) {
            return new TagParseResult({
                consumed: true,
                scope_open: {
                    type: 'if_builder',
                    branches: [],
                    current_branch_type: 'if',
                    current_branch_expr: ifm[1].trim(),
                    current_branch_steps: [],
                },
            });
        }

        const elifm = this.ELIF.exec(line);
        if (elifm) {
            const top = stack[stack.length - 1];
            if (!top || top.type !== 'if_builder') {
                throw new Error(
                    `Syntax Error line ${lineIdx + 1}: ` +
                    `Unexpected [elif] without an open [if].`
                );
            }
            if (top.current_branch_type === 'else') {
                throw new Error(
                    `Syntax Error line ${lineIdx + 1}: ` +
                    `[elif] cannot come after [else].`
                );
            }
            (top.branches as Branch[]).push({
                mode: top.current_branch_type as Branch['mode'],
                condition: top.current_branch_expr as string,
                steps: top.current_branch_steps as Step[],
            });
            top.current_branch_type = 'elif';
            top.current_branch_expr = elifm[1].trim();
            top.current_branch_steps = [];
            return new TagParseResult({ consumed: true });
        }

        if (this.ELSE.test(line)) {
            const top = stack[stack.length - 1];
            if (!top || top.type !== 'if_builder') {
                throw new Error(
                    `Syntax Error line ${lineIdx + 1}: ` +
                    `Unexpected [else] without an open [if].`
                );
            }
            if (top.current_branch_type === 'else') {
                throw new Error(
                    `Syntax Error line ${lineIdx + 1}: ` +
                    `Duplicate [else] in a single conditional block.`
                );
            }
            (top.branches as Branch[]).push({
                mode: top.current_branch_type as Branch['mode'],
                condition: top.current_branch_expr as string,
                steps: top.current_branch_steps as Step[],
            });
            top.current_branch_type = 'else';
            top.current_branch_expr = null;
            top.current_branch_steps = [];
            return new TagParseResult({ consumed: true });
        }

        if (this.CLOSE.test(line)) {
            const top = stack[stack.length - 1];
            if (!top || top.type !== 'if_builder') {
                throw new Error(`Syntax Error line ${lineIdx + 1}: Mismatched [/if].`);
            }
            return new TagParseResult({ consumed: true, scope_close: 'if_builder' });
        }

        return new TagParseResult({ consumed: false });
    }

    execute(step: Step, ctx: ExecContext): ExecResult {
        if (step.type !== 'conditional_block') return null;

        const branches = step.branches as Branch[];

        for (const branch of branches) {
            if (branch.mode === 'if' || branch.mode === 'elif') {
                const condition = branch.condition ?? 'false';
                let result: boolean;
                try {
                    result = evaluateCondition(condition, ctx.env);
                } catch (err) {
                    throw new Error(
                        `CONDITIONAL_EVAL_ERROR: '${condition}' — ${String(err)}`
                    );
                }
                if (result) return ['inject', branch.steps];
            } else {
                return ['inject', branch.steps];
            }
        }

        return null;
    }
}

function evaluateCondition(expr: string, env: Record<string, unknown>): boolean {
    const keys = Object.keys(env).filter(k => /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(k));
    const values = keys.map(k => env[k]);
    // eslint-disable-next-line @typescript-eslint/no-implied-eval
    const fn = new Function(...keys, `return (${expr});`);
    return Boolean(fn(...values));
}